gobblin-modules/gobblin-helix/src/main/java/gobblin/metastore/ZkStateStore.java

```diff
@@ -0,0 +1,279 @@
+  private final HelixPropertyStore<byte[]> propStore;
+  private final boolean compressedValues;
+
+  /**
+   * State store that stores instances of {@link State}s in a ZooKeeper-backed {@link HelixPropertyStore}
+   * @param connectString ZooKeeper connect string
+   * @param storeRootDir The root directory for the state store
+   * @param compressedValues should values be compressed for storage?
+   * @param stateClass The type of state being stored
+   * @throws IOException
+   */
+  public ZkStateStore(String connectString, String storeRootDir, boolean compressedValues, Class<T> stateClass)
+      throws IOException {
+    this.compressedValues = compressedValues;
+    this.stateClass = stateClass;
+
+    ZkSerializer serializer = new ByteArraySerializer();
+    propStore = new ZkHelixPropertyStore<byte[]>(connectString, serializer, storeRootDir);
+  }
+
+  private String formPath(String storeName) {
+    return "/" + storeName;
+  }
+
+  private String formPath(String storeName, String tableName) {
+    return "/" + storeName + "/" + tableName;
+  }
+
+  @Override
+  public boolean create(String storeName) throws IOException {
+    String path = formPath(storeName);
+
+    return propStore.exists(path, 0) || propStore.create(path, ArrayUtils.EMPTY_BYTE_ARRAY,
+        AccessOption.PERSISTENT);
+  }
+
+  @Override
+  public boolean create(String storeName, String tableName) throws IOException {
+    String path = formPath(storeName, tableName);
+
+    if (propStore.exists(path, 0)) {
+      throw new IOException(String.format("State already exists for storeName %s tableName %s", storeName,
+          tableName));
+    }
+
+    return propStore.create(path, ArrayUtils.EMPTY_BYTE_ARRAY, AccessOption.PERSISTENT);
+  }
+
+  @Override
+  public boolean exists(String storeName, String tableName) throws IOException {
+    String path = formPath(storeName, tableName);
+
+    return propStore.exists(path, 0);
+  }
+
+  private void addStateToOs(DataOutput dataOutput, T state) throws IOException {
+    new Text(Strings.nullToEmpty(state.getId())).write(dataOutput);
+    state.write(dataOutput);
+  }
+
+  @Override
+  public void put(String storeName, String tableName, T state) throws IOException {
+    putAll(storeName, tableName, Collections.singletonList(state));
+  }
+
+  @Override
+  public void putAll(String storeName, String tableName, Collection<T> states) throws IOException {
+    String path = formPath(storeName, tableName);
+    ByteArrayOutputStream byteArrayOs = new ByteArrayOutputStream();
+    OutputStream os = compressedValues ? new GZIPOutputStream(byteArrayOs) : byteArrayOs;
+    DataOutputStream dataOutput = new DataOutputStream(os);
+
+    for (T state : states) {
+      addStateToOs(dataOutput, state);
+    }
+
+    if (!exists(storeName, tableName) && !create(storeName, tableName)) {
+      throw new IOException("Failed to create a state file for table " + tableName);
+    }
+
+    dataOutput.close();
```

Comment:
The close here is to close and flush the compression stream so that the byte array in the put() is complete.  In case of exception we don't need the close.